File: src/ios/GWSdkBLEwrapper.m
//
//  GWSdkBLEwrapper.m
//  giz
//
//  Created by 陈东东 on 15/11/5.
//  Copyright © 2015年 陈东东. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <GizDataAccess/GizDataAccess.h>
#import <Cordova/CDV.h>

@interface GWSdkBLEwrapper: CDVPlugin<GizDataAccessSourceDelegate, GizDataAccessLoginDelegate> {
    // Member variables go here.
    NSString * _appId;
    GizDataAccessSource* gdaSource ;
    GizDataAccessLogin* gdaLogin;
}
-(void)login:(CDVInvokedUrlCommand *)command;
-(void)writeData:(CDVInvokedUrlCommand *)command;
-(void)readData:(CDVInvokedUrlCommand *)command;
@property (strong,nonatomic) CDVInvokedUrlCommand * commandHolder;

@end

@implementation GWSdkBLEwrapper

-(void)pluginInitialize:(CDVInvokedUrlCommand *) command{
    NSString *appid=command.arguments[0];
    if (appid!=nil) {
    if (!_appId) {
        [GizDataAccess startWithAppID:_appId];
     }
    }else {
        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"appid is null!"];
          [self.commandDelegate sendPluginResult:pluginResult callbackId:self.commandHolder.callbackId];
    }
    //Init SDK
    
    
    gdaSource = [[GizDataAccessSource alloc] initWithDelegate:self];
    gdaLogin = [[GizDataAccessLogin alloc] initWithDelegate:self];
    
}

-(void)login:(CDVInvokedUrlCommand *)command
{
    self.commandHolder = command;
    [self pluginInitialize:command];
    
    NSString *username=command.arguments[1];
    NSString *password=command.arguments[2];
    if(username&&password){
         [gdaLogin login:username password:password];
    }else{
        [gdaLogin loginAnonymous];
    }
   
}

- (void)gizDataAccessDidLogin:(GizDataAccessLogin *)login uid:(NSString *)uid token:(NSString *)token result:(GizDataAccessErrorCode)result message:(NSString *)message {
    if(result == kGizDataAccessErrorNone) {
        // 登录成功
        // ……
        NSLog(@"登录成功");
        NSDictionary *d = [NSDictionary dictionaryWithObjectsAndKeys:
                           @"登陆成功", @"state",
                           nil];
        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:d];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:self.commandHolder.callbackId];
    }
}

/**
 支持批量数据上传。最后一个参数data为数组类型，可以指定每一组数据的产生时间和内容，格式为标准的JSON。上传结果，通过数据上传的委托返回。
 
 data示例： @{@"ts": [timestamp], @"attrs": @{ [dynamic_keys]: [dynamic_values], ... } }
 
 */
-(void)writeData:(CDVInvokedUrlCommand *)command{
    NSString *token=command.arguments[1];
    NSString *productKey=command.arguments[2];
    NSString *deviceSn=command.arguments[3];
    NSArray *data=command.arguments[4];
    self.commandHolder = command;
    [self pluginInitialize:command];
    [gdaSource saveData:token productKey:productKey deviceSN:deviceSn data:data];
}

- (void)gizDataAccessDidSaveData:(GizDataAccessSource *)source result:(GizDataAccessErrorCode)result message:(NSString *)message {
    if(result == kGizDataAccessErrorNone) {
        // 数据上传成功
        // ……
        NSDictionary *d = [NSDictionary dictionaryWithObjectsAndKeys:
                           @"数据上传成功", @"state",
                           message,@"message",
                           nil];
        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:d];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:self.commandHolder.callbackId];
    }else{
        
        NSDictionary *d = [NSDictionary dictionaryWithObjectsAndKeys:
                           @"数据上传失败", @"state",
                           message,@"message",
                           nil];
        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsDictionary:d];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:self.commandHolder.callbackId];
    }
}

//获取数据时，需指定起止时间段。如果limit值为0将只返回20条数据，若skip值为负数，则获取失败。数据获取结果，通过获取数据的委托返回。获取到的数据，按照时间排序，最新的数据排在最前面。

-(void)readData:(CDVInvokedUrlCommand *)command{
    self.commandHolder = command;
    [self pluginInitialize:command];
    
    NSString *token=command.arguments[1];
    NSString *productKey=command.arguments[2];
    NSString *deviceSn=command.arguments[3];
    int16_t startTime=command.arguments[4];
    int16_t endTime=command.arguments[5];
    
    int  limit=20;
    int  skip=0;
    if(command.arguments[6]!=nil||command.arguments[6]!=0){
        limit=command.arguments[6];
    }
    if(command.arguments[7]!=nil){
        limit=command.arguments[7];
    }

    [gdaSource loadData:token productKey:productKey deviceSN:deviceSn startTime:startTime endTime:endTime limit:limit skip:skip];
    
    
}
/*实现获取数据的委托方法
 
 获取到的数据内容，按以下键值对方式提供：
 
 @"attrs" : { [dynamic_keys] : [dynamic_values], ... },
 
 @"uid" : [uid],
 
 @"sn" : [sn],
 
 @"ts" : [ts],
 
 @"product_key" : [product_key]
 */
- (void)gizDataAccessDidLoadData:(GizDataAccessSource *)source data:(NSArray *)data result:(GizDataAccessErrorCode)result errorMessage:(NSString *)message {
    if(result == kGizDataAccessErrorNone) {
        // 获取数据成功
        for(NSDictionary *dict in data) {
            NSString* sn = [dict valueForKey:@"device_sn"];
            NSString* productkey = [dict valueForKey:@"product_key"];
            NSString* uid = [dict valueForKey:@"uid"];
            NSNumber *nTS = [dict valueForKey:@"ts"];
            NSDictionary* attributes = [dict valueForKey:@"attrs"];
            
            NSLog(@"sn:%@ productkey:%@ uid:%@ nTS:%@ attributes:%@", sn, productkey, uid, nTS, attributes);
        }
        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsArray:data];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:self.commandHolder.callbackId];
    }else{
        NSDictionary *d = [NSDictionary dictionaryWithObjectsAndKeys:
                           @"数据获取失败", @"state",
                           message,@"message",
                           nil];
        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsDictionary:d];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:self.commandHolder.callbackId];
    }
}
@end

File: src/android/GWSdkBLEwrapper.java
package com.heytz.GWSdkBLEwrapper;


import android.content.Context;
import com.gizwits.gizdataaccess.GizDataAccess;
import com.gizwits.gizdataaccess.GizDataAccessLogin;
import com.gizwits.gizdataaccess.GizDataAccessSource;
import com.gizwits.gizdataaccess.entity.GizDataAccessErrorCode;
import com.gizwits.gizdataaccess.listener.GizDataAccessLoginListener;
import com.gizwits.gizdataaccess.listener.GizDataAccessSourceListener;
import org.apache.cordova.*;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by chendongdong on 15/11/8.
 */
public class GWSdkBLEwrapper extends CordovaPlugin {

    private CallbackContext cordovaCallbackContext;
    private Context context;
    private String _appId;
    private GizDataAccessLogin gdalogin;
    private GizDataAccessSource gdaSource;

    public GizDataAccessLoginListener gizDataAccessLoginListener = new GizDataAccessLoginListener() {
        public void didLogin(String uid, String token, GizDataAccessErrorCode result, String message) {
            if (result.getResult() == 0 && uid != null && token != null) {
                JSONObject json = new JSONObject();
                try {
                    json.put("uid", uid);
                    json.put("token", token);
                    // 登录成功
                    // ……
                    PluginResult pr = new PluginResult(PluginResult.Status.OK, json);
                    cordovaCallbackContext.sendPluginResult(pr);
                } catch (JSONException e) {
                    //e.printStackTrace();
                    PluginResult pr = new PluginResult(PluginResult.Status.ERROR, e.getMessage());
                    cordovaCallbackContext.sendPluginResult(pr);
                }
            } else {
                PluginResult pr = new PluginResult(PluginResult.Status.ERROR, message);
                cordovaCallbackContext.sendPluginResult(pr);
            }
        }
    };

    public GizDataAccessSourceListener gizDataAccessSourceListener = new GizDataAccessSourceListener() {
        public void didSaveData(GizDataAccessSource source, GizDataAccessErrorCode result, String message) {
            if (result.getResult() == 0) {
                // 上传成功
                // ……
                PluginResult pr = new PluginResult(PluginResult.Status.OK, "success");
                cordovaCallbackContext.sendPluginResult(pr);
            }
        }

        public void didLoadData(GizDataAccessSource arg0, JSONArray jsonArray, GizDataAccessErrorCode result, String message) {
            if (result.getResult() == 0) {
                if (jsonArray != null) {
                    for (int i = 0; i < jsonArray.length(); i++) {
                        try {
                            JSONObject jsonObject = jsonArray.getJSONObject(i);
                            System.out.println("uid: " + jsonObject.get("uid"));
                            System.out.println("device_sn: " + jsonObject.get("device_sn"));
                            System.out.println("product_key: " + jsonObject.get("product_key"));
                            System.out.println("ts: " + jsonObject.get("ts"));
                            System.out.println("attrs: " + jsonObject.get("attrs"));
                            System.out.println("\n");
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    }
                    PluginResult pr = new PluginResult(PluginResult.Status.OK, jsonArray);
                    cordovaCallbackContext.sendPluginResult(pr);
                } else {
                    System.out.println("暂无数据");
                }
            } else {
                System.out.println("读取失败：" + message);
            }
        }
    };

    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);
        // your init code here
        context = cordova.getActivity().getApplicationContext();
        gdalogin = new GizDataAccessLogin(gizDataAccessLoginListener);
        gdaSource = new GizDataAccessSource(gizDataAccessSourceListener);
    }


    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        cordovaCallbackContext = callbackContext;
        /**
         * 判断是否设置了appId
         */
        if (args.toString(0) != null && !args.toString(0).isEmpty()) {
            if (_appId == null || (_appId.compareToIgnoreCase(args.getString(0)) != 0)) {
                _appId = args.getString(0);
                GizDataAccess.startWithAppId(context, _appId);
            }
        } else {
            PluginResult pr = new PluginResult(PluginResult.Status.ERROR, "appId is null");
            cordovaCallbackContext.sendPluginResult(pr);
            return false;
        }
        /**
         * 登录的方法
         */
        if (action.equals("login")) {
            this.login(args.getString(1), args.getString(2), callbackContext);
            return true;
        }
        if (action.equals("writeData")) {
            List<String> dataList = new ArrayList<String>();
            JSONArray jsonArray = args.getJSONArray(4);

            for (int i = 0; i < jsonArray.length(); i++) {
                dataList.add(jsonArray.get(i).toString());
            }
            this.writeData(args.toString(1), args.toString(2), args.toString(3), dataList, callbackContext);
            return true;
        }
        if (action.equals("readData")) {
            this.readData(args.toString(1), args.toString(2), args.toString(3),
                    Long.parseLong(args.toString(4)),
                    Long.parseLong(args.toString(5)),
                    Integer.parseInt(args.toString(6)),
                    Integer.parseInt(args.toString(7)),
                    callbackContext);
            return true;
        }
        return false;
    }

    /**
     * 如果存在用户名和密码，那么使用这个用户名和密码登录，如果不存在那么使用匿名登录
     *
     * @param username        可选
     * @param password        可选
     * @param callbackContext
     */
    public void login(String username, String password, CallbackContext callbackContext) {

        if (username != null  && password != null &&username != "null"  && password != "null"&& (!username.isEmpty()) && (!password.isEmpty())) {
            gdalogin.login(username, password);
        } else {
            gdalogin.loginAnonymous();
        }
    }

    /**
     * 发送控制命令
     *
     * @param token
     * @param product_key
     * @param device_sn
     * @param data
     * @param callbackContext
     */
    public void writeData(String token, String product_key, String device_sn, List<String> data, CallbackContext callbackContext) {
        gdaSource.saveData(token, product_key, device_sn, data);
    }

    /**
     * @param token
     * @param product_key
     * @param device_sn
     * @param start_time
     * @param end_time
     * @param limit           可选，如果为null 或者 0 默认等于 20
     * @param skip            可选 如果等于 null 默认为 0;
     * @param callbackContext
     */
    public void readData(String token, String product_key, String device_sn, long start_time, long end_time, Integer limit, Integer skip, CallbackContext callbackContext) {
        if (limit == null || limit == 0) {
            limit = 20;
        }
        if (skip == null) {
            skip = 0;
        }
        gdaSource.loadData(token, product_key, device_sn, start_time, end_time, limit, skip);

    }
}
